web_services/foxx/brightid5/operations.js

```diff
@@ -0,0 +1,238 @@
+  }
+  if (!nacl.sign.detached.verify(strToUint8Array(message), b64ToUint8Array(sig), b64ToUint8Array(app.sponsorPublicKey))) {
+    throw 'invalid signature';
+  }
+}
+
+const senderAttrs = {
+  'Add Connection': ['id1', 'id2'],
+  'Remove Connection': ['id1'],
+  'Add Group': ['id1'],
+  'Remove Group': ['id'],
+  'Add Membership': ['id'],
+  'Remove Membership': ['id'],
+  'Set Trusted Connections': ['id'],
+  'Set Signing Key': ['id'],
+  'Sponsor': ['app'],
+  'Link ContextId': ['id'],
+  'Invite': ['inviter'],
+  'Dismiss': ['dismisser'],
+  'Add Admin': ['id'],
+};
+let operationsCount = {};
+let resetTime = 0;
+function checkLimits(op, timeWindow, limit) {
+  if (Date.now() > resetTime) {
+    operationsCount = {};
+    resetTime =  Date.now() + timeWindow;
+  }
+  const senders = senderAttrs[op.name].map(attr => op[attr]);
+  const usersColl = arango._collection('users');
+  for (let sender of senders) {
+    // these condition structure is applying:
+    // 1) a bucket for a verified user
+    // 2) a bucket for children of a verified user
+    // 3) a bucket for all non-verified users without parent
+    // where parent is the first verified user that make connection with the user
+    if (!usersColl.exists(sender)) {
+      // this happens when operation is "Add Connection" and one/both sides don't exist
+      sender = 'shared';
+    } else {
+      const user = usersColl.document(sender);
+      verified = user.verifications && user.verifications.includes('BrightID');
+      if (!verified && user.parent) {
+        // this happens when user is not verified but has a verified connection
+        sender = `shared_${user.parent}`;
+      } else if (!verified && !user.parent) {
+        // this happens when user is not verified and does not have a verified connection
+        sender = 'shared';
+      }
+    }
+    if (!operationsCount[sender]) {
+      operationsCount[sender] = 0;
+    }
+    operationsCount[sender] += 1;
+    if (operationsCount[sender] <= limit) {
+      // if operation has multiple senders, this check will be passed
+      // even if one of the senders did not reach limit yet
+      return;
+    }
+  }
+  throw 'Too Many Requests';
+}
+
+function verify(op) {
+  if (op.v != 5) {
+    throw 'invalid operation version';
+  }
+  if (op.timestamp > Date.now() + TIME_FUDGE) {
+    throw "timestamp can't be in the future";
+  }
+  const requiredSigs = {
+    'Add Connection': [['id1', 'sig1'], ['id2', 'sig2']],
+    'Remove Connection': [['id1', 'sig1']],
+    'Add Group': [['id1', 'sig1']],
+    'Remove Group': [['id', 'sig']],
+    'Add Membership': [['id', 'sig']],
+    'Remove Membership': [['id', 'sig']],
+    'Set Trusted Connections': [['id', 'sig']],
+    'Set Signing Key': [['id1', 'sig1'], ['id2', 'sig2']],
+    'Sponsor': [['app', 'sig']],
+    'Link ContextId': [['id', 'sig']],
+    'Invite': [['inviter', 'sig']],
+    'Dismiss': [['dismisser', 'sig']],
+    'Add Admin': [['id', 'sig']],
+  }
```

Comment:
initialize this constant outside of the function.

Reply:
applied :pray: